stdlib/public/core/Diffing.swift

```diff
@@ -0,0 +1,805 @@
+      let change: Change
+      if enumeratedRemoves < removals.count && enumeratedInserts < insertions.count {
+        let removeOffset = removals[enumeratedRemoves]._offset
+        let insertOffset = insertions[enumeratedInserts]._offset
+        if removeOffset - enumeratedRemoves <= insertOffset - enumeratedInserts {
+          change = removals[enumeratedRemoves]
+        } else {
+          change = insertions[enumeratedInserts]
+        }
+      } else if enumeratedRemoves < totalRemoves {
+        change = removals[enumeratedRemoves]
+      } else if enumeratedInserts < totalInserts {
+        change = insertions[enumeratedInserts]
+      } else {
+        // Not reached, loop should have exited.
+        preconditionFailure()
+      }
+
+      consume(change)
+
+      switch change {
+      case .remove(_, _, _):
+        enumeratedRemoves += 1
+      case .insert(_, _, _):
+        enumeratedInserts += 1
+      }
+    }
+  }
+}
+
+extension RangeReplaceableCollection {
+  /// Applies a difference to a collection.
+  ///
+  /// - Parameter difference: The difference to be applied.
+  ///
+  /// - Returns: An instance representing the state of the receiver with the
+  ///   difference applied, or `nil` if the difference is incompatible with
+  ///   the receiver's state.
+  ///
+  /// - Complexity: O(*n* + *c*), where *n* is `self.count` and *c* is the
+  ///   number of changes contained by the parameter.
+  @available(macOS 9999, iOS 9999, tvOS 9999, watchOS 9999, *) // FIXME(availability-5.1)
+  public func applying(_ difference: CollectionDifference<Element>) -> Self? {
+    var result = Self()
+    var enumeratedRemoves = 0
+    var enumeratedInserts = 0
+    var enumeratedOriginals = 0
+    var currentIndex = self.startIndex
+
+    func append(
+      into target: inout Self,
+      contentsOf source: Self,
+      from index: inout Self.Index, count: Int) {
+      let start = index
+      source.formIndex(&index, offsetBy: count)
+      target.append(contentsOf: source[start..<index])
+    }
+
+    difference._fastEnumeratedApply { change in
+      switch change {
+      case .remove(offset: let offset, element: _, associatedWith: _):
+        let origCount = offset - enumeratedOriginals
+        append(into: &result, contentsOf: self, from: &currentIndex, count: origCount)
+        enumeratedOriginals += origCount + 1 // Removal consumes an original element
+        currentIndex = self.index(after: currentIndex)
+        enumeratedRemoves += 1
+      case .insert(offset: let offset, element: let element, associatedWith: _):
+        let origCount = (offset + enumeratedRemoves - enumeratedInserts) - enumeratedOriginals
+        append(into: &result, contentsOf: self, from: &currentIndex, count: origCount)
+        result.append(element)
+        enumeratedOriginals += origCount
+        enumeratedInserts += 1
+      }
+      assert(enumeratedOriginals <= self.count)
```

Comment:
`assert` is rarely the right choice for the stdlib code. Consider using `_precondition`, `_debugPrecondition`, or `_internalInvariant` as described [here](https://github.com/apple/swift/blob/master/docs/StandardLibraryProgrammersManual.md#precondition).